src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -282,6 +283,39 @@ impl KvmVcpu {
         Ok(cpuid)
     }
 
+    /// If the IA32_TSC_DEADLINE MSR value is zero, update it
+    /// with the IA32_TSC value to guarantee that
+    /// the vCPU will continue receiving interrupts after restoring from a snapshot.
+    ///
+    /// Rationale: we observed that sometimes when taking a snapshot,
+    /// the IA32_TSC_DEADLINE MSR is cleared, but the interrupt is not
+    /// delivered to the guest, leading to a situation where one
+    /// of the vCPUs never receives TSC interrupts after restoring,
+    /// until the MSR is updated externally, eg by setting the system time.
+    fn fix_zero_tsc_deadline_msr(msr_chunks: &mut [Msrs]) {
+        // We do not expect more than 1 TSC MSR entry, but if there are multiple, pick the maximum.
+        let max_tsc_value = msr_chunks
+            .iter()
+            .flat_map(|msrs| msrs.as_slice())
+            .filter(|msr| msr.index == MSR_IA32_TSC)
+            .map(|msr| msr.data)
+            .max();
+
+        if let Some(tsc_value) = max_tsc_value {
+            for msrs in msr_chunks.iter_mut() {
+                for msr in msrs.as_mut_slice().iter_mut() {
+                    if msr.index == MSR_IA32_TSC_DEADLINE && msr.data == 0 {
+                        warn!(
+                            "MSR_IA32_TSC_DEADLINE is 0, replacing with {:x}.",
+                            tsc_value
+                        );
+                        msr.data = tsc_value;
+                    }
+                }
+            }
+        }
```

Comment:
Here is a version with a bit less indentation.
```Rust
let Some(tsc_value) = msr_chunks
            .iter()
            .flat_map(|msrs| msrs.as_slice())
            .filter(|msr| msr.index == MSR_IA32_TSC)
            .map(|msr| msr.data) else { continue; };

for msr in msr_chunks
            .iter_mut()
            .flat_map(|msrs| msrs.as_mut_slice())
            .filter(|msr| msr.index == MSR_IA32_TSC_DEADLINE && msr.data == 0) {
  warn!(
    "MSR_IA32_TSC_DEADLINE is 0, replacing with {:x}.",
    tsc_value
  );
  msr.data = tsc_msr.data;
}
```

Reply:
Not sure how `continue` can be used outside a loop. Could you elaborate?
I got rid of the nested loop, let me know if it looks better to you.

Reply:
My mistake, there should have been `return` instead of `continue`.
I think your current version is fine.